caproto/_log.py

```diff
@@ -143,10 +155,128 @@ def color_logs(color):
 
 
 logger = logging.getLogger('caproto')
+ch_logger = logging.getLogger('caproto.ch')
+search_logger = logging.getLogger('caproto.bcast.search')
 current_handler = None  # overwritten below
 
 
-def set_handler(file=sys.stdout, datefmt='%H:%M:%S', color=True):
+class LoggerNameFilter(logging.Filter):
+    def __init__(self, logger_names, level='NOSET'):
+        self.logger_names = logger_names
+        self.levelno = validate_level(level)
+    def filter(self, record):
+        for i in self.logger_names:
+            if fnmatch.fnmatch(record.name, i) and record.levelno >= self.levelno:
+                return True
+        return False
+
+
+def validate_level(level)-> int:
+    '''
+    Return a int for level comparison
+
+    '''
+    if isinstance(level, int):
+        levelno = level
+    elif isinstance(level, str):
+        levelno = logging.getLevelName(level)
+
+    if isinstance(levelno, int):
+        return levelno
+    else:
+        raise CaprotoValueError("The target addresses should given as a list of strings "
+                "like 'XX.XX.XX.XX:YYYY' "
+                "or tuples like ('XX.XX.XX.XX', YYYY).")
+
+class PVFilter(logging.Filter):
+    '''
+    Pass through only messages relevant to one or more PV names and
+    pv-unrelated messages.
+
+    Parameters
+    ----------
+    names : string or list of string
+        PV name or PV name list which will be filtered in.
+
+    Returns
+    -------
+    Bool: True or False
+        True if message is not PV related which doesn't have 'pv' as key in extra
+        True if 'pv' as key exists and pv name exists in Filter list.
+        False if message is PV related but pv name isn't in Filter list.
+    '''
+    def __init__(self, names, level='NOTSET', exclusive=False):
+        self.names = names
+        self.levelno = validate_level(level)
+        self.exclusive = exclusive
+
+    def filter(self, record):
```

Comment:
I think we agreed on this behavior:

```python
if record.levelno >= self.levelno and not self.exclusive:
    # The record exceeds filter's barrier height and can pass regardless
    # of whether it satisfies the condition.
    return True
for i in self.names:
    if  fnmatch.fnmatch(record.pv, i):
        # The record satisfies the condition.
        return True
return False
```

which is not equivalent to the latest version in this PR. For example, in the agreed-on version a record with a `levelno` below `self.levelno` will pass if it satisfies the condition. In the version in this PR, it will not pass. Literally copy-pasting this agreed-upon version, including the comments, would be fine by me unless you think there need to be changes.

Reply:
Per conversation on phone / Slack, both Ke and I have the wrong solution, but we now agree on the right one, and better yet we have tests, added in 48d03089.